src/Psl/Hash/Context.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Hash;
+
+use HashContext;
+use Psl;
+use Psl\Arr;
+
+use function hash_final;
+use function hash_init;
+use function hash_update;
+
+/**
+ * Incremental hashing context.
+ *
+ * Example:
+ *
+ *      Hash\Context::create('md5')
+ *          ->update('The quick brown fox ')
+ *          ->update('jumped over the lazy dog.')
+ *          ->finalize()
+ *      => Str("5c6ffbdd40d9556b73a21e63c3e0e904")
+ *
+ * @psalm-immutable
+ */
+final class Context
+{
+    private HashContext $internalContext;
+
+    private function __construct(HashContext $internal_context)
+    {
+        $this->internalContext = $internal_context;
+    }
+
+    /**
+     * Initialize an incremental hashing context.
+     *
+     * @throws Psl\Exception\InvariantViolationException If the given algorithm is unsupported.
+     */
+    public static function create(string $algorithm): Context
+    {
+        Psl\invariant(Arr\contains(algorithms(), $algorithm), 'Unsupported algorithm "%s".', $algorithm);
+        $internal_context = hash_init($algorithm);
+
+        return new self($internal_context);
+    }
+
+    /**
+     * Pump data into an active hashing context.
+     */
+    public function update(string $data): Context
+    {
+        /** @var HashContext $internal_context */
+        $internal_context = hash_copy($this->internalContext);
+        hash_update($internal_context, $data);
+
+        return new self($internal_context);
+    }
+
+    /**
+     * Finalize an incremental hash and return resulting digest.
+     */
+    public function finalize(): string
+    {
+        /** @var HashContext $internal_context */
+        $internal_context = hash_copy($this->internalContext);
+
+        return hash_final($internal_context, false);
+    }
```

Comment:
@veewee what do you think about this?

Finalizing the context here doesn't free the internal hash context, since this object is immutable, it makes sense from the type system prespactive, you can be sure that passing this object around will never result in any modifications, however, but i'm not really sure 🤔 with that method name, it might be excpected that the context is now freed from the memory.

do you think we should free the internal context at this stage? should we rename this method?

Reply:
Not sure ...

Since you copy the internal context first, the final call will always succeed.
I don't think you should free the context, because that will result in an unusable Context class.

Because it is immutable, you can finalize all intermediate context states as well.
From that point of view, it would make sense not to nuke the internal context resource either.

On the other hand, there is also the memory question from above that comes to mind.


I am not really sure what the best approach is at this moment.
Maybe we can make a descision based on the anser to the memory question?

Reply:
Assuming the context will most likely only live inside a function that returns the hash string, I wouldn't care too much about freeing up the resource.

Reply:
exactly my thought.

I have updated `hash` functions to use context as well ( avoding `hash`/`hash_hmac` builtin functions ).